File: models/attendance/guard_duty_board.py
# -*- coding: utf-8 -*-

from collections import defaultdict

from odoo import api, fields, models

from ..shared.schedule_report_mixin import HOUR_EPSILON

WEEKDAYS = ('0', '1', '2', '3', '4')
# Mirrors ems.group's own SHIFT_HOURS (models/contacts/group_schedule.py) — the guard duty board
# needs the same morning/afternoon split, but rendered as weekday x shift tables with one column
# per group rather than one column per weekday, so it can't reuse that method as-is.
SHIFT_HOURS = {
    'morning': (8, 15),
    'afternoon': (15, 22),
}

# Which absence requests are worth putting on the board, and how each one is reported to the
# screen. 'approved' is a fact to plan around; 'pending' is a warning that one may be coming -
# hr_holidays' two pre-approval states are deliberately collapsed into a single one here,
# because the difference between "waiting for the first approver" and "waiting for the second"
# changes nothing for whoever is assigning guards. Refused and cancelled requests are absent
# from this mapping entirely, which is what keeps them off the board.
ABSENCE_STATES = {
    'confirm': 'pending',
    'validate1': 'pending',
    'validate': 'approved',
}

# A whole-day absence, as an hour interval, so it can be compared against a schedule period the
# same way a partial one is - no special case anywhere downstream.
WHOLE_DAY = (0.0, 24.0)


def _period_contains(container, period):
    """True when 'period' (hour_from, hour_to) is fully covered by 'container' - same
    containment check '_merge_absorbed_periods' below needs inline, extracted so the level
    filter's guard/break folding (issue #390) can reuse it against a period that isn't
    necessarily a member of the same 'periods' list (see 'get_guard_duty_board_lines' docstring).
    Uses HOUR_EPSILON for the same float-noise reason as '_merge_absorbed_periods'."""
    return container[0] <= period[0] + HOUR_EPSILON and container[1] >= period[1] - HOUR_EPSILON


def _merge_absorbed_periods(periods):
    """Groups 'periods' (a list of distinct (hour_from, hour_to) tuples) into
    {period: [period, *periods absorbed into it]}, one entry per period that keeps its own row -
    an absorbed period is never a key of the returned dict.

    A period is absorbed into another one from the same list when its own hour range is fully
    contained in the other's (found 2026-09 as issue #410: a teacher's own personal schedule can
    end a guard-duty slot early - e.g. a shorter working day that block - while a colleague's
    guard for the same start time runs the full period; the same thing happens for any other
    non-teaching activity, like a 35-minute coordination duty sitting next to a 60-minute class.
    Before this, get_guard_duty_board_lines() rendered the short period as its own row, almost or
    entirely empty). If several periods could contain a given one, the smallest (shortest) is
    preferred, to avoid folding into an unnecessarily wide row should several levels of
    containment ever coexist.

    Uses HOUR_EPSILON (see its own NOTE) since two periods that are conceptually "the same" can
    still differ by a hair's-width float remainder depending on how each was computed."""
    containers = {}
    for period in periods:
        candidates = [
            other for other in periods
            if other != period and _period_contains(other, period)
        ]
        if candidates:
            containers[period] = min(candidates, key=lambda other: other[1] - other[0])

    def root(period):
        # 'seen' guards against a cycle between two periods that both fall within HOUR_EPSILON of
        # each other in both bounds (so each looks like a - vanishingly narrow - "container" of
        # the other) - not a real containment relation, just float noise; without this a cycle
        # would loop root() forever instead of just resolving to whichever period was reached
        # first.
        seen = set()
        while period in containers and period not in seen:
            seen.add(period)
            period = containers[period]
        return period

    members = {}
    for period in periods:
        members.setdefault(root(period), []).append(period)
    return members


class EmsCourseGuardDutyBoard(models.Model):
    # NOTE: extends 'ems.course' in place (not a new model) — a 2-item '_inherit' list without an
    # explicit '_name' would make Odoo's metaclass define a brand-new model instead (see MetaModel
    # in odoo/models.py). Same pattern as ems_working_schedule/resource.calendar and
    # ems_group_schedule/ems.group.
    #
    # This lives on 'ems.course' (a real, always-existing, already-readable-by-every-teacher
    # model — see security/ir.model.access.csv's ems.access_ems_course_teacher) rather than a
    # dedicated TransientModel wizard, deliberately: an earlier version used a TransientModel
    # opened via a dynamic ir.actions.server, which meant the URL bar showed a raw
    # "ems.guard_duty_board/<id>" instead of a stable "action-<xmlid>" like every other EMS
    # screen — Odoo can only put a real xmlid in the URL for a *statically declared* action, and
    # a server action that returns a dynamically-built act_window dict has no xmlid of its own to
    # show. Binding to 'ems.course' instead means the board's own screen is a plain
    # ir.actions.client (a real, static, URL-addressable action — see
    # views/attendance/guard_duty_board/menu.xml) with no per-visit record to create/open at all.
    _name = 'ems.course'
    _inherit = ['ems.course', 'ems.schedule_report_mixin']

    def _get_guard_duty_board_attendance_ids(self):
        """Every real (non-framework) teacher's Mon-Fri attendance row, across every teacher —
        same aggregation idea as ems.group._compute_schedule_attendance_ids
        (models/contacts/group_schedule.py), generalized from "this group" to "the whole centre".
        Deliberately not filtered by 'calendar_id.course_id' — mirrors that same precedent, which
        aggregates the same way without a course filter either: a course-transitioned-out
        calendar/attendance row is archived (active=False, see
        ems_working_schedule_assignation.active's own NOTE), so this plain search() already only
        ever returns the current course's real, active schedules, and stays correct even for a
        legacy calendar whose 'course_id' was never backfilled (added 2026-08-06, not every
        pre-existing row necessarily has it set).

        The explicit 'calendar_id.active' check below is defense-in-depth, not redundant with the
        above: 'ems_working_schedule.action_archive()' now cascades to every remaining attendance
        row when a calendar itself is retired, but this search must not silently start trusting
        that invariant everywhere it's ever established - a calendar could in principle end up
        archived by some other path without its own attendance rows following (found 2026-09-01:
        before that cascade existed, a rolled-over teacher's ARCHIVED calendar kept showing
        active non-teaching rows here indefinitely, since a bare 'active=True' row search never
        looks at its own parent calendar's active state at all).

        'calendar_id.employee_id != False' excludes any calendar that isn't a specific teacher's
        own personal working schedule - concretely, Odoo's own generic default calendar (e.g.
        "Standard 40 hours/week", auto-created with its own Mon-Fri 8-12/13-17 attendance rows
        the first time anything needs 'res.company.resource_calendar_id' and nothing has been
        customized yet). That calendar is never itself is_framework=True, so the check above
        alone doesn't exclude it, and on a clean install with no real schedules configured yet
        it silently shows up here and corrupts period computation (found 2026-09-08 via CI: the
        generic 8-12 block contains/absorbs a real, narrower test period into itself)."""
        return self.env['resource.calendar.attendance'].search([
            ('calendar_id.is_framework', '=', False),
            ('calendar_id.active', '=', True),
            ('calendar_id.employee_id', '!=', False),
            ('dayofweek', 'in', WEEKDAYS),
        ])

    def _get_guard_duty_absence_intervals(self, day, employees):
        """`{employee.id: [(hour_from, hour_to, state)]}` - the absences covering `day`.

        Resolved on demand rather than stored anywhere: an absence is an 'hr.leave' keyed to a
        real date, the board is keyed to a weekday, and nothing joins the two until a date is
        actually asked for. Returns `{}` for no date at all, which is what lets the PDF (and any
        other weekday-only caller) keep working unchanged.

        sudo() because a guard-duty absence is not private to its own approval chain: whoever
        reads this board legitimately needs to know that a colleague is not coming, which is
        exactly the same justification 'get_guard_sessions()' carries for reading schedules that
        are not the reader's own. Only the fact and the interval are ever exposed - never the
        absence type, its reason, or its attachments.
        """
        if not day or not employees:
            return {}
        leaves = self.env['hr.leave'].sudo().search([
            ('employee_id', 'in', employees.ids),
            ('state', 'in', list(ABSENCE_STATES)),
            ('request_date_from', '<=', day),
            ('request_date_to', '>=', day),
        ])
        intervals = defaultdict(list)
        for leave in leaves:
            # 'request_unit_hours' rather than 'ems_full_day': it is the field that actually
            # decides whether request_hour_from/to carry anything (see hr.leave's own
            # _compute_request_unit_hours and EMS's override of it), and a multi-day request is
            # a whole day on each of its days regardless of how it was filled in.
            partial = leave.request_unit_hours and leave.request_date_from == leave.request_date_to
            hours = (leave.request_hour_from, leave.request_hour_to) if partial else WHOLE_DAY
            intervals[leave.employee_id.id].append((*hours, ABSENCE_STATES[leave.state]))
        return intervals

    @staticmethod
    def _guard_duty_absence_state(intervals, employees, hour_from, hour_to):
        """`{employee.id: 'approved'|'pending'}` for those of `employees` absent in the period.

        An approved absence outranks a pending one when the same teacher has both overlapping
        the same period: the period is going to need covering either way, so reporting it as
        merely "requested" would understate it.
        """
        states = {}
        for employee in employees:
            overlapping = {state for start, stop, state in intervals.get(employee.id, ())
                           if start < hour_to and stop > hour_from}
            if overlapping:
                states[employee.id] = 'approved' if 'approved' in overlapping else 'pending'
        return states

    def get_guard_duty_board_lines(self, weekday, shift, level_ids=None, day=None):
        """Board rows for one weekday + shift: the ordered list of group columns actually taught in
        that slot, one row per distinct time period (chronological), each with one cell per group
        (teacher(s) + room, or empty) plus the guard-duty teacher(s) for that period. A guard row
        carries no group of its own (see 'non_teaching'), so it's reported separately from the
        group columns instead of as one more column. No per-subject colour is computed here (an
        earlier version reused ems.schedule_report_mixin's REPORT_COLOR_PALETTE the same way the
        teacher/group schedule PDFs do) - removed per developer feedback (2026-09-01): with every
        group already its own column and the subject spelled out as a short acronym in the cell,
        a colour-per-subject wash added visual noise without adding information a plain table
        didn't already convey.

        A period whose hour range is fully absorbed by another period starting at the same time
        (or earlier) never gets a row of its own - see _merge_absorbed_periods() - its cells/
        guards are folded into the row of the period that contains it instead (issue #410).

        A period left with no teaching cell AND no guard after that merge (e.g. a coordination
        duty/meeting nobody's actual class or guard shift overlaps) is dropped entirely - a row
        with nothing in it but a time range adds no information (developer follow-up on #410,
        2026-09-07).

        'level_ids' (issue #390, falsy/omitted = "All levels", the previous, still-default
        behaviour): a list of 'ems.level' ids to narrow the board down to. See "Level filter" in
        docs/en/developers/attendance/guard_duty_board.md for the full design. When set:
        - 'teaching_entries'/'groups'/'periods' only ever come from a group whose own 'level_id'
          (see ems.group's own field) is in 'level_ids' - a "reinforcement" group (no level_id by
          design) can never match. This is the only thing the filter actually controls: which
          time blocks (rows) exist at all.
        - A guard is shown on whichever ROW its own period overlaps, full stop - there is no way
          to know, and no need to know, which level a guard-duty teacher "belongs" to (developer
          feedback, 2026-09-07, after an earlier version tried deriving it from the teacher's own
          other classes that day and got it backwards for a guard who simply doesn't teach
          anything that day - see [[project_guard_duty_board_level_filter]] in memory): once a
          time block is visible for this level (because some class of it runs then), every guard
          on duty then is relevant, regardless of what they otherwise teach.
        - A guard whose own period isn't covered by any of the rows above still gets a shot at a
          dedicated "Patio" row if it falls inside that level's own break period - see
          _get_guard_duty_board_break_lines(). One left over after that simply has no visible
          block to attach to under this level and isn't shown (still visible under "All levels").

        Checking every existing level in the client's own filter dropdown is normalized to the
        same "All levels" (falsy) path below, not treated as a real filter: even with the guard
        rule above, a level-filtered view only ever builds rows from teaching entries (never from
        a guard-only or "reinforcement"-group period the way the unfiltered path's own 'entries'-
        wide 'periods' does) - checking literally every level is meant to mean "show everything",
        indistinguishable from checking none at all.

        `day`, when given, is the concrete date the weekday stands for, and is what adds the
        absence information: every cell gains an 'absences' map of which of its own teachers are
        away, each row gains 'guard_absences' for the same question asked of the guard column,
        and 'absences' - the list of classes actually left without a teacher, which is what the
        board's second tab is built from. Without a date none of that can be resolved at all, so
        every one of those comes back empty and the board is the plain timetable it was before."""
        self.ensure_one()
        if level_ids and set(level_ids) >= set(self.env['ems.level'].search([]).ids):
            level_ids = None
        day = fields.Date.to_date(day)
        shift_start, shift_end = SHIFT_HOURS[shift]
        entries = self._get_guard_duty_board_attendance_ids().filtered(
            lambda attendance: attendance.dayofweek == weekday
                and attendance.hour_from >= shift_start and attendance.hour_to <= shift_end)
        teaching_entries = entries.filtered(lambda attendance: attendance.subject_id or attendance.group_ids)
        guard_entries = entries.filtered(lambda attendance: attendance.non_teaching_is_guard)
        intervals = self._get_guard_duty_absence_intervals(day, entries.employee_id)

        if level_ids:
            teaching_entries = teaching_entries.filtered(
                lambda attendance: attendance.group_ids.filtered(lambda group: group.level_id.id in level_ids))
            groups = teaching_entries.group_ids.filtered(lambda group: group.level_id.id in level_ids).sorted(key=lambda group: group.name)
            periods = sorted({(attendance.hour_from, attendance.hour_to) for attendance in teaching_entries})
        else:
            groups = teaching_entries.group_ids.sorted(key=lambda group: group.name)
            periods = sorted({(attendance.hour_from, attendance.hour_to) for attendance in entries})
        period_members = _merge_absorbed_periods(periods)

        dated_lines = []
        remaining_guards = guard_entries
        for hour_from, hour_to in periods:
            if (hour_from, hour_to) not in period_members:
                continue  # absorbed - already folded into its container's row below
            member_periods = period_members[(hour_from, hour_to)]
            cells = []
            covering = []
            for group in groups:
                cell_entries = teaching_entries.filtered(
                    lambda attendance, group=group, member_periods=member_periods:
                        group in attendance.group_ids and (attendance.hour_from, attendance.hour_to) in member_periods
                )
                # Every co-teacher for this cell, deduped (a plain recordset union already
                # does that) - a co-taught slot has one 'resource.calendar.attendance' row
                # per teacher, all sharing the same group/period, so 'entries' alone would
                # silently drop every name but the first one picked for display.
                cell_teachers = cell_entries.mapped('employee_id')
                cell_absences = self._guard_duty_absence_state(
                    intervals, cell_teachers, hour_from, hour_to)
                cells.append({
                    'group': group,
                    'entries': cell_entries,
                    'teachers': cell_teachers,
                    'absences': cell_absences,
                })
                # One row per absent teacher AND per class of theirs in this period: a teacher
                # splitting the period across two groups leaves two classes uncovered, and each
                # one has to be assigned its own guard.
                first = cell_entries[:1]
                covering += [{
                    'teacher': teacher,
                    'state': cell_absences[teacher.id],
                    'group': group,
                    'subject': first.subject_id,
                    'room': first.space_id,
                } for teacher in cell_teachers if teacher.id in cell_absences]
            if level_ids:
                # A guard's own period is no longer necessarily one of 'periods' above (those now
                # only come from teaching_entries) - fold it into whichever row's range genuinely
                # contains it, not just an exact-tuple match.
                guards = remaining_guards.filtered(
                    lambda attendance, hour_from=hour_from, hour_to=hour_to:
                        _period_contains((hour_from, hour_to), (attendance.hour_from, attendance.hour_to)))
            else:
                guards = guard_entries.filtered(
                    lambda attendance, member_periods=member_periods:
                        (attendance.hour_from, attendance.hour_to) in member_periods)
            remaining_guards -= guards
            if not guards and not any(cell['entries'] for cell in cells):
                continue  # nothing scheduled anywhere in this period - no row worth showing
            dated_lines.append((hour_from, hour_to, {
                'time_label': "%s-%s" % (self._format_report_time(hour_from), self._format_report_time(hour_to)),
                'cells': cells,
                'guards': guards,
                # Not folded into 'absences' below: an absent guard has no class of their own for
                # anyone to cover, they are simply one fewer person available to cover somebody
                # else's - a subtraction from the guard column, not an addition to the work.
                'guard_absences': self._guard_duty_absence_state(
                    intervals, guards.mapped('employee_id'), hour_from, hour_to),
                'absences': covering,
            }))

        if level_ids:
            dated_lines += self._get_guard_duty_board_break_lines(
                level_ids, weekday, shift_start, shift_end, groups, remaining_guards, intervals)
            dated_lines.sort(key=lambda dated_line: (dated_line[0], dated_line[1]))
        return {'groups': groups, 'lines': [line for _hour_from, _hour_to, line in dated_lines]}

    def _get_guard_duty_board_break_lines(self, level_ids, weekday, shift_start, shift_end, groups, unmatched_guards, intervals):
        """Once a level filter is active, that level's own break ("Patio") period has no real
        teaching entry of its own to build a row from (a teacher's own calendar spans one
        continuous block across it - see hr.employee._get_derived_break_entries' own docstring),
        so a guard duty scheduled specifically for a break would otherwise simply disappear once
        'get_guard_duty_board_lines' restricts its rows to the filtered level's teaching periods
        only. Any leftover 'unmatched_guards' entry (any guard not already folded into a row
        above - see the caller, which no longer derives a guard's own level at all) that falls
        inside one of that level's own framework(s)' break periods gets a dedicated row instead -
        marked 'is_break' for the client to label distinctly. A
        break period with no guard inside it renders nothing, same "nothing to show" rule every
        other row already follows. Not meaningful under "All levels" - a break's own hours differ
        per level (see docs/en/developers/attendance/guard_duty_board.md's own "Level filter"
        section), so the caller only ever invokes this once a level filter narrows down which
        break(s) apply.

        'intervals' is forwarded straight from the caller so a guard's own absence still shows up
        on their dedicated break row exactly like it would on any other row - see
        get_guard_duty_board_lines()'s own 'guard_absences'."""
        frameworks = self.env['resource.calendar'].search([
            ('is_framework', '=', True), ('level_id', 'in', level_ids),
        ])
        if not frameworks:
            return []
        break_periods = sorted({
            (attendance.hour_from, attendance.hour_to)
            for attendance in self.env['resource.calendar.attendance'].search([
                ('calendar_id', 'in', frameworks.ids),
                ('dayofweek', '=', weekday),
                ('non_teaching_is_break', '=', True),
                ('hour_from', '>=', shift_start), ('hour_to', '<=', shift_end),
            ])
        })
        empty_entries = self.env['resource.calendar.attendance']
        dated_lines = []
        for hour_from, hour_to in break_periods:
            guards = unmatched_guards.filtered(
                lambda attendance, hour_from=hour_from, hour_to=hour_to:
                    _period_contains((hour_from, hour_to), (attendance.hour_from, attendance.hour_to)))
            if not guards:
                continue  # nothing to show for this break slot
            unmatched_guards -= guards
            dated_lines.append((hour_from, hour_to, {
                'time_label': "%s-%s" % (self._format_report_time(hour_from), self._format_report_time(hour_to)),
                'cells': [{
                    'group': group, 'entries': empty_entries, 'teachers': empty_entries.employee_id, 'absences': {},
                } for group in groups],
                'guards': guards,
                'guard_absences': self._guard_duty_absence_state(
                    intervals, guards.mapped('employee_id'), hour_from, hour_to),
                'absences': [],
                'is_break': True,
            }))
        return dated_lines

    @api.model
    def get_guard_duty_board_data(self, weekday, shift, level_ids=None, day=None):
        """JSON-safe wrapper around get_guard_duty_board_lines(), for the guard duty board client
        action's own RPC call (static/src/js/backend/guard_duty_board.js). @api.model: resolves
        "the current course" itself (env.company.current_course_id), so the JS side never needs
        to know or pass a specific ems.course id — matches how the aggregation itself is scoped
        (see _get_guard_duty_board_attendance_ids' own NOTE: not actually course-filtered).
        'level_ids' (issue #390) and 'day' are forwarded as-is - see get_guard_duty_board_lines()'s
        own docstring. Every teacher is reported as {'name', 'absence'} rather than a bare name,
        so both of the screen's tabs read absences the same way, off the same payload, instead of
        the client having to match names back against a separate list."""
        course = self.env.company.get_current_course_or_raise()
        data = course.get_guard_duty_board_lines(weekday, shift, level_ids=level_ids, day=day)
        groups = [{'id': group.id, 'name': group.name} for group in data['groups']]
        lines = []
        for line in data['lines']:
            cells = []
            for cell in line['cells']:
                first = cell['entries'][:1]
                # 'acronym' (e.g. "MP 0440"), not 'display_name' (which also spells out the full
                # subject name) - the cell is tight on space, the full name is one click away on
                # the teacher's own schedule.
                cells.append({
                    'group_id': cell['group'].id,
                    'subject': first.subject_id.acronym if first else False,
                    'teachers': self._guard_duty_teacher_data(cell['teachers'], cell['absences']),
                    'room': first.space_id.display_name if first and first.space_id else False,
                })
            lines.append({
                'time_label': line['time_label'],
                'cells': cells,
                'guards': self._guard_duty_teacher_data(
                    line['guards'].mapped('employee_id'), line['guard_absences']),
                'absences': [{
                    'teacher': row['teacher'].display_name,
                    'state': row['state'],
                    'group': row['group'].name,
                    'subject': row['subject'].acronym if row['subject'] else False,
                    'room': row['room'].display_name if row['room'] else False,
                } for row in line['absences']],
                'is_break': line.get('is_break', False),
            })
        return {'groups': groups, 'lines': lines}

    @staticmethod
    def _guard_duty_teacher_data(teachers, absences):
        """Teachers as `[{'name', 'absence'}]` - 'absence' being False, 'approved' or 'pending'."""
        return [{'name': teacher.display_name, 'absence': absences.get(teacher.id, False)}
                for teacher in teachers]

    @api.model
    def get_guard_duty_board_levels(self):
        """Every 'ems.level', for the guard duty board's own level filter (issue #390) - plain
        id/name pairs, JSON-safe like get_guard_duty_board_data(). No course scoping: a level is
        centre-wide curriculum data, not tied to any one ems.course."""
        return [{'id': level.id, 'name': level.name} for level in self.env['ems.level'].search([])]

    @api.model
    def get_current_course_data(self):
        """Small helper for the guard duty board's client action: the page has no bound record of
        its own to read 'the current course' from (see the class docstring above for why), so it
        asks for it explicitly instead — used both to label the page and to supply the PDF
        button's own 'active_ids'."""
        course = self.env.company.get_current_course_or_raise()
        return {'id': course.id, 'name': course.name}

File: static/src/js/backend/guard_duty_board.js
/** @odoo-module **/

import { Component, onWillStart, useState } from "@odoo/owl";
import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { _t } from "@web/core/l10n/translation";
import { dayLabels } from "./schedule_grid_geometry";

const SHIFTS = [
    { key: "morning", label: _t("Morning") },
    { key: "afternoon", label: _t("Afternoon") },
];

// The two ways of reading the same day: the timetable everyone already knows, with whoever is
// away struck through it, and the plain "who is missing / who is on guard" list built from the
// very same payload (see ems.course.get_guard_duty_board_data) - one fetch, two renderings.
const VIEWS = [
    { key: "schedule", label: _t("Guard duty schedule") },
    { key: "table", label: _t("Absences table") },
];

const MS_PER_DAY = 24 * 60 * 60 * 1000;

// Mirrors ems.course's own SHIFT_HOURS (models/attendance/guard_duty_board.py): the afternoon
// shift starts at 15:00, kept in sync by hand since a plain JS constant can't share the Python
// one directly.
const AFTERNOON_START_HOUR = 15;

// Defaults the board to whichever day/shift the viewer would actually want to see right now
// (developer feedback, 2026-09-01: "hoy es martes... como ya son las 15h, pues el turno de
// tarde") — the browser's own local clock, since "now" here means the viewer's own wall-clock
// time, not the server's. Date.getDay() is 0=Sunday..6=Saturday; our own day index is 0=Monday.
// ..4=Friday, so a weekend falls outside that range - Monday is as reasonable a fallback as any
// (the board has no "weekend" concept at all, every table is keyed to a Mon-Fri dayofweek).
function getDefaultDayAndShift() {
    const now = new Date();
    const jsDay = now.getDay();
    const day = jsDay >= 1 && jsDay <= 5 ? jsDay - 1 : 0;
    const shift = now.getHours() >= AFTERNOON_START_HOUR ? "afternoon" : "morning";
    return { day, shift };
}

// "YYYY-MM-DD" from the browser's own local calendar date - deliberately NOT toISOString(),
// which converts to UTC first and so hands back the previous day for anyone east of Greenwich
// during the evening. The board's date is a plain calendar day, never an instant in time.
function toIsoDate(date) {
    const month = String(date.getMonth() + 1).padStart(2, "0");
    const day = String(date.getDate()).padStart(2, "0");
    return `${date.getFullYear()}-${month}-${day}`;
}

function fromIsoDate(iso) {
    const [year, month, day] = iso.split("-").map(Number);
    return new Date(year, month - 1, day);
}

// The Monday of whichever week `date` falls in. A weekend date belongs to the week that just
// happened (Saturday and Sunday follow their own Monday), which is also what makes picking a
// weekend in the date input land on a real, showable weekday instead of nothing at all.
function mondayOf(date) {
    const jsDay = date.getDay();
    const offset = jsDay === 0 ? 6 : jsDay - 1;
    return new Date(date.getTime() - offset * MS_PER_DAY);
}

// Read-only, centre-wide board: one weekday visible at a time (tabs), morning/afternoon are
// different shifts (never merged into one table — see ems.group.shift) picked via a dropdown
// within the active day, not shown stacked together — columns are the groups actually taught in
// that shift, rows are time blocks, plus the guard-duty teacher(s) for each block. A guard slot
// has no group of its own (see resource.calendar.attendance.group_ids' own NOTE on non-teaching
// rows), so guards are reported per-row instead of as one more column.
//
// A plain ir.actions.client (registered below), not a form view on a wizard record — see
// models/attendance/guard_duty_board.py's own class docstring for why: a wizard record's URL
// would show a raw "model/id" instead of a proper "action-<xmlid>" like every other EMS screen.
//
// Data is fetched via RPC, one weekday/shift at a time (ems.course.get_guard_duty_board_data()) —
// not read from any field's own prefetched sub-records, unlike the teacher/group/student grids
// (schedule_grid_field.js/schedule_grid_readonly_field.js). Those aggregate at most one teacher's,
// one group's or one student's own schedule; this one aggregates the whole centre (easily several hundred rows),
// which the web client's own x2many sub-record fetch silently caps — an earlier version that did
// read a prefetched field this way only ever showed real data for whichever weekday loaded first.
export class GuardDutyBoard extends Component {
    static template = "ems.GuardDutyBoard";
    static props = ["*"];

    setup() {
        this.orm = useService("orm");
        this.actionService = useService("action");
        const { day, shift } = getDefaultDayAndShift();
        this.state = useState({
            activeDay: day,
            activeShift: shift,
            // Issue #390's level filter: empty = "All levels" (the previous, still-default
            // behaviour) - see guard_duty_board.py's own get_guard_duty_board_lines() docstring.
            activeLevelIds: [],
            levels: [],
            // The Monday the weekday tabs hang off. Absences are keyed to real dates while the
            // timetable is keyed to weekdays, so the board needs a concrete week before it can
            // say who is away - see ems.course.get_guard_duty_board_lines()'s 'day' argument.
            weekStart: toIsoDate(mondayOf(new Date())),
            activeView: "schedule",
            board: null,
            loading: true,
            courseId: null,
            courseName: "",
        });
        onWillStart(async () => {
            const [course, levels] = await Promise.all([
                this.orm.call("ems.course", "get_current_course_data", []),
                this.orm.call("ems.course", "get_guard_duty_board_levels", []),
            ]);
            this.state.courseId = course.id;
            this.state.courseName = course.name;
            this.state.levels = levels;
            await this.loadBoard();
        });
    }

    // Each weekday tab carries its own real date, so the tab strip doubles as the week's
    // calendar: the label says which Monday, not just "Monday".
    get days() {
        const monday = fromIsoDate(this.state.weekStart);
        return dayLabels().map((label, index) => {
            const date = new Date(monday.getTime() + index * MS_PER_DAY);
            return { index, label, date: toIsoDate(date), dayOfMonth: date.getDate() };
        });
    }

    // The header's own title - a plain string literal in the template would never go through
    // the translation extractor at all (found 2026-09-11: it always rendered in English
    // regardless of the user's own language, unlike every other label on this screen).
    get title() {
        return this.state.courseName
            ? _t("Guard duty schedule (%s)", this.state.courseName)
            : _t("Guard duty schedule");
    }

    get shifts() {
        return SHIFTS;
    }

    get views() {
        return VIEWS;
    }

    // The date the board is actually showing: the active weekday tab within the active week.
    get activeDate() {
        return this.days[this.state.activeDay].date;
    }

    get columnLabels() {
        return {
            time: _t("Time block"),
            guard: _t("Guard duty"),
            absences: _t("Absences"),
        };
    }

    get emptyLabels() {
        return {
            schedule: _t("No schedule for this shift."),
            absences: _t("Nobody is missing this shift."),
            loading: _t("Loading..."),
        };
    }

    // Compact label for the level dropdown's own toggle button - the full checkbox list already
    // shows every level by name, this is just what's visible before opening it.
    get levelFilterLabel() {
        if (!this.state.activeLevelIds.length) {
            return _t("All levels");
        }
        const selected = this.state.levels.filter((level) => this.state.activeLevelIds.includes(level.id));
        return selected.map((level) => level.name).join(", ");
    }

    async setActiveDay(index) {
        if (index === this.state.activeDay) {
            return;
        }
        this.state.activeDay = index;
        await this.loadBoard();
    }

    setActiveView(key) {
        // Both tabs render the same already-fetched payload, so switching between them never
        // costs a round trip.
        this.state.activeView = key;
    }

    async onShiftChange(ev) {
        const key = ev.target.value;
        if (key === this.state.activeShift) {
            return;
        }
        this.state.activeShift = key;
        await this.loadBoard();
    }

    async toggleLevel(levelId) {
        const activeLevelIds = this.state.activeLevelIds;
        const index = activeLevelIds.indexOf(levelId);
        if (index === -1) {
            activeLevelIds.push(levelId);
        } else {
            activeLevelIds.splice(index, 1);
        }
        await this.loadBoard();
    }

    // Picking any date moves the whole week and lands on that date's own weekday tab - a
    // weekend picks the week it closes (see mondayOf) and falls back to its Monday, since the
    // board has no weekend column to show.
    async onDateChange(ev) {
        const value = ev.target.value;
        if (!value) {
            return;
        }
        const picked = fromIsoDate(value);
        const monday = mondayOf(picked);
        const jsDay = picked.getDay();
        this.state.weekStart = toIsoDate(monday);
        this.state.activeDay = jsDay >= 1 && jsDay <= 5 ? jsDay - 1 : 0;
        await this.loadBoard();
    }

    async shiftWeek(weeks) {
        const monday = fromIsoDate(this.state.weekStart);
        this.state.weekStart = toIsoDate(new Date(monday.getTime() + weeks * 7 * MS_PER_DAY));
        await this.loadBoard();
    }

    async loadBoard() {
        this.state.loading = true;
        this.state.board = await this.orm.call(
            "ems.course",
            "get_guard_duty_board_data",
            [String(this.state.activeDay), this.state.activeShift, this.state.activeLevelIds, this.activeDate]
        );
        this.state.loading = false;
    }

    // Bold red for an absence that is going to happen, a lighter italic for one still waiting
    // on its approver - the planner has to be able to tell a fact from a warning at a glance.
    absenceClass(absence) {
        if (absence === "approved") {
            return "o_guard_board_absent";
        }
        return absence === "pending" ? "o_guard_board_absent_pending" : "";
    }

    // One PDF per day AND per shift — whichever day tab / shift dropdown is currently active,
    // not the whole week or both shifts — see reports/attendance/report_guard_duty_board.xml's
    // own use of the 'guard_duty_weekday'/'guard_duty_shift' context keys. 'guard_duty_level_ids'
    // (issue #390) forwards the same level selection, following the same pattern; 'guard_duty_date'
    // is the printed copy's own absence information - a cuadrante handed out to plan the day's
    // guards is no use without them. 'guard_duty_view' (found 2026-09-11) forwards which of the
    // two tabs is actually on screen, so the PDF prints whatever the user is currently looking
    // at instead of always the schedule tab regardless of the "Absences table" tab being active.
    async onPdfClick() {
        await this.actionService.doAction("ems.action_report_guard_duty_board", {
            additionalContext: {
                active_ids: [this.state.courseId],
                guard_duty_weekday: String(this.state.activeDay),
                guard_duty_shift: this.state.activeShift,
                guard_duty_level_ids: this.state.activeLevelIds,
                guard_duty_date: this.activeDate,
                guard_duty_view: this.state.activeView,
            },
        });
    }
}

registry.category("actions").add("ems_guard_duty_board", GuardDutyBoard);
